src/crimson/os/seastore/btree/fixed_kv_node.h

```diff
@@ -84,11 +84,55 @@ struct FixedKVNode : ChildableCachedExtent {
    * 	cannot be rewritten) because their parents must be mutated upon remapping.
    */
   std::vector<ChildableCachedExtent*> children;
-  std::set<FixedKVNodeRef, copy_source_cmp_t> copy_sources;
+  std::set<FixedKVNodeRef, fixedkv_node_cmp_t> copy_sources;
   uint16_t capacity = 0;
   parent_tracker_t* my_tracker = nullptr;
   RootBlockRef root_block;
 
+  // copy dests points from a stable node back to its pending nodes
+  // having copy sources at the same tree level, it serves as a two-level index:
+  // transaction-id then node-key to the pending node.
+  //
+  // The copy dest pointers must be symmetric to the copy source pointers.
+  //
+  // copy_dests_t will be automatically unregisterred upon transaction destruction,
+  // see Transaction::views
+  struct copy_dests_t : trans_spec_view_t {
+    std::set<FixedKVNodeRef, fixedkv_node_cmp_t> dests_by_key;
+    copy_dests_t(Transaction &t) : trans_spec_view_t{t.get_trans_id()} {}
+    ~copy_dests_t() {
+      LOG_PREFIX(~copy_dests_t);
+      SUBTRACE(seastore_fixedkv_tree, "copy_dests_t destroyed");
+    }
+  };
+
+  trans_view_set_t copy_dests_by_trans;
+
+  void add_copy_dest(Transaction &t, FixedKVNodeRef dest) {
+    ceph_assert(is_stable());
+    ceph_assert(dest->is_pending());
+    auto tid = t.get_trans_id();
+    auto iter = copy_dests_by_trans.lower_bound(
+      tid, trans_spec_view_t::cmp_t());
+    if (iter == copy_dests_by_trans.end() ||
+	iter->pending_for_transaction != tid) {
+      iter = copy_dests_by_trans.insert_before(
+	iter, t.add_transactional_view<copy_dests_t>(t));
+    }
+    auto &copy_dests = static_cast<copy_dests_t&>(*iter);
+    copy_dests.dests_by_key.insert(dest);
```

Comment:
Is it safer to assert that the insertion must be successful (insertion must took place) ?

Reply:
Um, actually, there seems to be cases in which duplicated insertions are inevitable, for example:

1. stable LBANode `L1` is split into two pending `L2` and `L3`
2. after a while, the same transaction merges `L2` and `L3` into `L4`
3. when pusing copy sources, the transaction may try to add `L4` into `L1`'s `dests_by_key` twice.

Reply:
Thanks for the scenario!

I see, how about implementing an assert that if `dest` cannot be inserted, the conflicting extent must be the `dest`.

Is that reasonable?

Reply:
yes, will do

Reply:
Done:-)